storage/remote/dial_context_test.go

```diff
@@ -0,0 +1,147 @@
+)
+
+type mockDialContext struct {
+	mock.Mock
+
+	addrFrequencyMu sync.Mutex
+	addrFrequency   map[string]int
+}
+
+func newMockDialContext(acceptableAddresses []string) *mockDialContext {
+	m := &mockDialContext{
+		addrFrequencyMu: sync.Mutex{},
+		addrFrequency:   make(map[string]int),
+	}
+	for _, acceptableAddr := range acceptableAddresses {
+		m.On("dialContext", mock.Anything, mock.Anything, acceptableAddr).Return(nil, nil)
+	}
+	return m
+}
+
+func (dc *mockDialContext) dialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	dc.addrFrequencyMu.Lock()
+	defer dc.addrFrequencyMu.Unlock()
+	args := dc.MethodCalled("dialContext", ctx, network, addr)
+	dc.addrFrequency[addr]++
+	return nil, args.Error(1)
+}
+
+func (dc *mockDialContext) getCount(addr string) int {
+	dc.addrFrequencyMu.Lock()
+	defer dc.addrFrequencyMu.Unlock()
+	return dc.addrFrequency[addr]
+}
+
+type mockedLookupHost struct {
+	withErr bool
+	result  []string
+}
+
+func (lh *mockedLookupHost) LookupHost(context.Context, string) ([]string, error) {
+	if lh.withErr {
+		return nil, errMockLookupHost
+	}
+	return lh.result, nil
+}
+
+func createDialContextWithRoundRobinDNS(dialContext config.DialContextFunc, resolver hostResolver) dialContextWithRoundRobinDNS {
+	return dialContextWithRoundRobinDNS{
+		dialContext: dialContext,
+		resolver:    resolver,
+	}
+}
+
+func TestDialContextWithRandomConnections(t *testing.T) {
+	numberOfRuns := 2 * len(testLookupResult)
+	var mdc *mockDialContext
+	testCases := map[string]struct {
+		addr  string
+		setup func() dialContextWithRoundRobinDNS
+		check func()
+	}{
+		"if address contains no port call default DealContext": {
+			addr: testAddrWithoutPort,
+			setup: func() dialContextWithRoundRobinDNS {
+				mdc = newMockDialContext([]string{testAddrWithoutPort})
+				return createDialContextWithRoundRobinDNS(mdc.dialContext, &mockedLookupHost{withErr: false})
+			},
+			check: func() {
+				require.Equal(t, numberOfRuns, mdc.getCount(testAddrWithoutPort))
+			},
+		},
+		"if lookup host returns error call default DealContext": {
+			addr: testAddrWithPort,
+			setup: func() dialContextWithRoundRobinDNS {
+				mdc = newMockDialContext([]string{testAddrWithPort})
+				return createDialContextWithRoundRobinDNS(mdc.dialContext, &mockedLookupHost{withErr: true})
+			},
+			check: func() {
+				require.Equal(t, numberOfRuns, mdc.getCount(testAddrWithPort))
+			},
+		},
+		"if lookup host is successful, shuffle results": {
+			addr: testAddrWithPort,
+			setup: func() dialContextWithRoundRobinDNS {
+				mdc = newMockDialContext(testLookupResultWithPort)
+				return createDialContextWithRoundRobinDNS(mdc.dialContext, &mockedLookupHost{result: testLookupResult})
+			},
+			check: func() {
+				// we ensure that not all runs will choose the first element of the lookup
+				require.NotEqual(t, numberOfRuns, mdc.getCount(testLookupResultWithPort[0]))
```

Comment:
Isn't this test flaky? I'd set the RNG seed for the test for determinism. 

For reference: I see the dialer is called 6 times, and there are 3 IPs. My math says there's a 1/3^5 chances this fails (1/243).